src/software/jetson_nano/services/motor.cpp

```diff
@@ -102,54 +111,223 @@ MotorService::MotorService(const RobotConstants_t& robot_constants,
+        static_cast<double>(back_left_rpm), static_cast<double>(back_right_rpm)};
+
+    TbotsProto::MotorStatus motor_status;
+    motor_status.mutable_front_right()->set_wheel_rpm(front_right_rpm);
+    motor_status.mutable_front_left()->set_wheel_rpm(front_left_rpm);
+    motor_status.mutable_back_left()->set_wheel_rpm(back_left_rpm);
+    motor_status.mutable_back_right()->set_wheel_rpm(back_right_rpm);
+
+    switch (motor.drive_control_case())
+    {
+        case TbotsProto::MotorControl::DriveControlCase::kDirectPerWheelControl:
+        {
             tmc4671_setTargetVelocity(
                 FRONT_LEFT_MOTOR_CHIP_SELECT,
+                static_cast<int>(motor.direct_per_wheel_control().front_left_wheel_rpm() *
+                                 robot_constants_.wheel_rotations_per_motor_rotation));
+            tmc4671_setTargetVelocity(
+                FRONT_RIGHT_MOTOR_CHIP_SELECT,
                 static_cast<int>(
-                    direct_control.direct_per_wheel_control().front_left_wheel_rpm() *
-                    RANDOM_SCALING_FACTOR));
-
+                    motor.direct_per_wheel_control().front_right_wheel_rpm() *
+                    robot_constants_.wheel_rotations_per_motor_rotation));
+            tmc4671_setTargetVelocity(
+                BACK_LEFT_MOTOR_CHIP_SELECT,
+                static_cast<int>(motor.direct_per_wheel_control().back_left_wheel_rpm() *
+                                 robot_constants_.wheel_rotations_per_motor_rotation));
+            tmc4671_setTargetVelocity(
+                BACK_RIGHT_MOTOR_CHIP_SELECT,
+                static_cast<int>(motor.direct_per_wheel_control().back_right_wheel_rpm() *
+                                 robot_constants_.wheel_rotations_per_motor_rotation));
             break;
         }
-        case TbotsProto::DirectControlPrimitive::WheelControlCase::kDirectVelocityControl:
+        case TbotsProto::MotorControl::DriveControlCase::kDirectVelocityControl:
         {
-            // TODO (#2335) convert local velocity to per-wheel velocity
-            // using http://robocup.mi.fu-berlin.de/buch/omnidrive.pdf and then
-            // communicate velocities to trinamic.
+            EuclideanSpace_t target_euclidean_velocity = {
+                motor.direct_velocity_control().velocity().x_component_meters(),
+                motor.direct_velocity_control().velocity().y_component_meters(),
+                motor.direct_velocity_control().angular_velocity().radians_per_second(),
+            };
+
+            // This is a linear transformation, we don't need to convert to/from
+            // RPM to MPS
+            WheelSpace_t target_speeds = euclidean_to_four_wheel.getTargetWheelSpeeds(
+                target_euclidean_velocity, current_wheel_speeds);
+
+            tmc4671_setTargetVelocity(FRONT_RIGHT_MOTOR_CHIP_SELECT,
+                                      static_cast<int>(target_speeds[0]));
```

Comment:
it seems like we're duplicating code with the above `tmc4671_setTargetVelocity`

Reply:
one of them is setting the rpms from the wheel conversion and the other one is from robot diagnostics